### PR TITLE
backend: WIP add llvmlite jit

### PR DESCRIPTION
**Pull request by positr0nium**
This PR introduces code that enables leveraging the llvmlit jit compiler for executing xDSL code. The function can be imported via ``from xdsl.backend.llvm import llvm_jit``. 

Example usage:

```python
from xdsl.dialects.builtin import ModuleOp, IntegerType
from xdsl.dialects.func import FuncOp, ReturnOp
from xdsl.ir.core import Region, Block

func_block = Block(arg_types = (IntegerType(64),))
func_block.add_op(ReturnOp(func_block.args[0]))

func_op = FuncOp(name = "test_function",
       function_type = ((IntegerType(64),), (IntegerType(64),)),
       region = Region(func_block)
       )

module_block = Block()
module_block.add_op(func_op)

module = ModuleOp(Region(module_block))
module.verify()

from xdsl.backend.llvm import llvm_jit

jitted_function = llvm_jit(module, "test_function")

print(jitted_function(42))
```

### CODE REVIEW

File: xdsl/xdsl_opt_main.py [hunk 1]
```diff
@@ -300,6 +301,37 @@ def _output_wgsl(prog: ModuleOp, output: IO[str]):
...
+                "--memref-expand",
+                "--fold-memref-alias-ops",
+                "--finalize-memref-to-llvm",
+                "--reconcile-unrealized-casts",
+                "-",
+            ]
+            mlir_opt = subprocess.Popen(
+                mlir_opt_passes, stdin=subprocess.PIPE, stdout=subprocess.PIPE
+            )
+            mlir_translate = subprocess.Popen(
```
Comment (superlopuh) [inline xdsl/xdsl_opt_main.py:323]: Sorry I probably should have been more clear. What I'm proposing to start with is essentially a reimplementation of this `mlir-translate` call. So no jitting to start with, just emission.

Comment (superlopuh) [reply xdsl/xdsl_opt_main.py:323]: The llvm backend target should print LLVM IR

Comment (positr0nium) [reply xdsl/xdsl_opt_main.py:323]: Ahh, okay - this might take some time though since I'd have some reading to do to achieve this.

Comment (superlopuh) [reply xdsl/xdsl_opt_main.py:323]: I'm happy to iterate with you here! Ping me on Zulip if you'd like any pointers I'm hoping it should be relatively straightforward (at most 30 mins of work)

File: xdsl/backend/llvm/__init__.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
+import ctypes
+import subprocess
+
+import llvmlite
```
Comment (superlopuh) [inline xdsl/backend/llvm/__init__.py:4]: Let's ignore this error for now
```suggestion
import llvmlite # pyright: ignore[reportMissingTypeStubs]
```

Comment (superlopuh) [reply xdsl/backend/llvm/__init__.py:4]: Similarly below

File: xdsl/backend/llvm/__init__.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        return ctypes.c_bool
+    if xdsl_type == Float32Type():
+        return ctypes.c_float
+    if xdsl_type == Float64Type():
+        return ctypes.c_double
+
+    raise TypeError(f"Unsupported or unknown xDSL type: {xdsl_type}")
+
+
+def xdsl_to_llvmlite_types(xdsl_type):
```
Comment (superlopuh) [inline xdsl/backend/llvm/__init__.py:38]: can you please add types here?

File: xdsl/xdsl_opt_main.py [hunk 4]
```diff
@@ -300,6 +301,37 @@ def _output_wgsl(prog: ModuleOp, output: IO[str]):
                     printer = WGSLPrinter(stream=output)
                     printer.print(op)
 
+        def _output_llvm(prog: ModuleOp, output: IO[str]):
+            mlir_text = str(prog)
```
Comment (superlopuh) [inline xdsl/xdsl_opt_main.py:305]: let's call the new helper here, and print it?

**Review by superlopuh — COMMENTED** [review]
Can you please add a new filecheck test that verifies that for an empty module a correctly formatted empty LLVM module is printed?

File: xdsl/backend/llvm/__init__.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                }
+
+                llvm_lite_block = llvm_lite_func.append_basic_block(name="entry")
+                translate_to_llvm_lite_block(
+                    xdsl_block, llvm_lite_block, ssa_translation
+                )
+
+            else:
+                raise Exception(
+                    "Tried to convert operation type {type(op} in top-level module to LLVM"
```
Comment (alexarice) [inline xdsl/backend/llvm/__init__.py:101]:
```suggestion
                    f"Tried to convert operation type {type(op)} in top-level module to LLVM"
```

**Comment by superlopuh** [comment]
Hi @positr0nium, I hope you're having a nice winter break! @suezli told me that he has some cycles to look into making a minimal version of this in the next few days, which will hopefully make it easier to work together towards the LLVM JIT as we discussed. Hope that's OK with you, and please let me know when you'd like to come back to this as there's more than enough work for the three of us in this area :)

**Comment by positr0nium** [comment]
Hey there! Thanks, I just came back from a not-so-wintery break in Thailand 🌴 I might find some time between the years but I'll be definitely back for it in the new year! Hope you had some nice festive days 😊

**Comment by superlopuh** [comment]
Thank you for opening this and iterating with me! I recently added a small test/demo of how I think this could work once we build up a bit more infrastructure: #5943.

I think we can safely close this PR for now, and instead slowly migrate bits of the prototype to the xDSL codebase.
